spec/expression.dd

```diff
@@ -28,6 +28,52 @@ Note: Although the definition is straightforward, a few subtleties exist related
 statement `return (() => x + f())() * g();`, the full expression of `f()` is `x + f()`, not the expression passed
 to `return`. This is because the parent of `x + f()` has function literal type, not expression type.
 
+$(DDOC_ANCHOR define-rvalue)$(P $(B Definition) ($(DOUBLEQUOTE Rvalue)): The following expressions,
+and no others, are called rvalue expressions or rvalues:)
+$(OL
+    $(LI `null`, `true`, `false`, `$`, `this`, `super`;)
+    $(LI literals: integral, floating point, character, string, array, associative array, function;)
+    $(LI `enum` values;)
+    $(LI the result of the following expressions:)
+    $(UL
+        $(LI functions that return by value, whether explicitly invoked or as the result of lowering
+        user-defined operators;)
+        $(LI the ternary operator `?:` when the two choices have different types, or at least one is an 
+        rvalue;)
+        $(LI built-in binary operators `||`, `&&`, `|`, `^`, `&`, `==`, `!=`, `is`, `!is`, `<`, `<=`, `>`,
+        `>=`, `in`, `<<`, `>>`, `+`, `-`, `~`, `*`, `/`, `%`, `^^`;)
+        $(LI built-in unary operators `&` (yields a pointer by value), `++` (postfix only), `--` (postfix
+        only), `-`, `!`, `~`, `cast`, `[]` (only when used for dynamic subslicing, e.g. `e[]` or 
+        `e[e1 .. e2]`), `new`;)
+        $(LI `mixin` expressions if and only if the compilation of the expression resulting from compiling
+        the argument(s) to `mixin` is an rvalue;)
+        $(LI `import` expressions;)
+        $(LI `typeid`;)
+        $(LI `is` expressions;)
+        $(LI the special keywords `__FILE__`, `__FILE_FULL_PATH__`, `__MODULE__`, `__LINE__`,
+        `__FUNCTION__`, `__PRETTY_FUNCTION__`, `__DATE__`, `__TIME__`, `__TIMESTAMP__`, `__VENDOR__`,
+        `__VERSION__`.)))
+
+$(DDOC_ANCHOR define-lvalue)$(P $(B Definition) ($(DOUBLEQUOTE Lvalue)): The following expressions, and no
+others, are called lvalue expressions or lvalues:)
+$(OL
+$(LI variable names;)
+$(LI function names;)
+$(LI the result of the $(I DotIdentifier) grammatical construct `.` when the rightmost side of the dot
+is the name of a field (whether direct or `static`);)
+$(LI the result of the following expressions:
+$(UL
+$(LI built-in unary operators `+`, `*`, `++` (prefix only), `--` (prefix only);)
+$(LI built-in binary operators `[]`, `=`, `+=`, `*=`, `/=`, `%=`, `&=`, `|=`, `^=`, `~=`,
+`<<=`, `>>=`, `>>>=`, `^^=`;)
+$(LI the ternary operator `?:` when the two choices have identical types and are lvalues;)
```

Comment:
does `alias this` play any tricksies here?

Reply:
Hrm, it actually does and in a not-super-consistent manner. Will update.